File: day05/02/02_1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strings"
)

var minV = 999999999

func match(s1, s2 string) bool {
	return strings.ToLower(s1) == strings.ToLower(s2) && s1 != s2
}

func produce(str string) string {
	l := len(str)
	if l == 0 {
		return ""
	} else if l == 1 {
		return str
	} else {
		half := int(l / 2)

		return merge(
			produce(str[:half]),
			produce(str[half:]),
		)
	}
}

func merge(front, back string) string {
	if front == "" && back == "" {
		return ""
	} else if front == "" {
		return back
	} else if back == "" {
		return front
	} else if match(front[len(front)-1:], back[0:1]) {
		return merge(front[:len(front)-1], back[1:])
	} else {
		return front + back
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func producer(s string) int {
	return len(produce(s))
}

func main() {
	file, _ := os.Open("input.txt")
	// file, _ := os.Open("sample.txt")

	defer file.Close()

	scanner := bufio.NewScanner(file)

	var str string

	for scanner.Scan() {
		str = scanner.Text()
	}

	m := make(map[string]bool)

	for _, char := range str {
		m[strings.ToLower(string(char))] = true
	}

	strList := make([]string, 0, len(m))
	for k := range m {
		r, _ := regexp.Compile("([" + strings.ToUpper(k) + k + "])")
		s := r.ReplaceAllString(str, "")
		strList = append(strList, s)
	}
	chInt := make(chan int)

	go func() {
		for _, s := range strList {
			chInt <- producer(s)
		}
		close(chInt)
	}()

	for l := range chInt {
		minV = min(l, minV)
	}
	fmt.Println(minV)

}

File: day05/02_1.js
const fs = require('fs')

function match(s1, s2) {
  return s1.toLowerCase() == s2.toLowerCase() && s1 !== s2
}

function produce(str) {
  const l = str.length

  if (l === 0) {
    return ''
  } else if (l === 1) {
    return str
  } else {
    const half = Math.round(l / 2)

    const left = str.slice(0, half)
    const right = str.slice(half)

    return merge(produce(left), produce(right))
  }
}

function merge(left, right) {
  if (left === '' && right === '') {
    return ''
  } else if (left === '') {
    return right
  } else if (right === '') {
    return left
  } else if (match(left.charAt(left.length - 1), right.charAt(0))) {
    return merge(left.slice(0, left.length - 1), right.slice(1))
  } else {
    return left + right
  }
}

function run() {
  // const data = fs.readFileSync('sample.txt', 'utf8')
  const data = fs.readFileSync('input.txt', 'utf8')

  const m = {}

  for (let i = 0; i < data.length; i++) {
    m[data[i].toLowerCase()] = true
  }

  const strList = Object.keys(m).map((ch) => {
    const regex = new RegExp(`([${ch.toUpperCase()}${ch}])`, 'g')
    return data.replace(regex, '')
  })

  let min = 99999999

  for (let i = 0; i < strList.length; i++) {
    str = strList[i]
    min = Math.min(produce(str).length, min)
  }

  console.log(min)
}

run()